Runtime/Retry/RetryHandler.cs

```diff
@@ -0,0 +1,142 @@
+﻿using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using Cysharp.Threading.Tasks;
+using Extreal.Core.Common.System;
+using Extreal.Core.Logging;
+using UniRx;
+
+namespace Extreal.Core.Common.Retry
+{
+    public class RetryHandler<TResult> : DisposableBase
+    {
+        [SuppressMessage("Usage", "CC0033")]
+        private readonly CompositeDisposable disposables = new CompositeDisposable();
+
+        public IObservable<int> OnRetrying => onRetrying.AddTo(disposables);
+        [SuppressMessage("Usage", "CC0033")] private readonly Subject<int> onRetrying = new Subject<int>();
+
+        public IObservable<bool> OnRetried => onRetried.AddTo(disposables);
+        [SuppressMessage("Usage", "CC0033")] private readonly Subject<bool> onRetried = new Subject<bool>();
+
+        private static readonly ELogger Logger = LoggingManager.GetLogger(nameof(RetryHandler<TResult>));
+
+        private readonly Func<UniTask<TResult>> runAsync;
+        private readonly Func<Exception, bool> isRetryable;
+        private readonly IRetryStrategy retryStrategy;
+
+        private static readonly IRetryStrategy NullRetryStrategy = new NullRetryStrategy();
+
+        private RetryHandler(
+            Func<UniTask<TResult>> runAsync, Func<Exception, bool> isRetryable, IRetryStrategy retryStrategy = null)
+        {
+            this.runAsync = runAsync ?? throw new ArgumentNullException(nameof(runAsync));
+            this.isRetryable = isRetryable ?? throw new ArgumentNullException(nameof(isRetryable));
+            this.retryStrategy = retryStrategy ?? NullRetryStrategy;
+        }
+
+        public async UniTask<TResult> HandleAsync(CancellationToken cancellationToken = default)
+        {
+            var retryCount = 0;
+            retryStrategy.Reset();
+            while (true)
+            {
+                try
+                {
+                    var result = await runAsync.Invoke();
+                    onRetried.OnNext(true);
+                    return result;
+                }
+                catch (Exception e)
+                {
+                    LogException(e);
+                    if (retryStrategy.HasNext() && isRetryable(e))
+                    {
+                        try
+                        {
+                            await UniTask.Delay(retryStrategy.Next(), cancellationToken: cancellationToken);
+                        }
+                        catch (OperationCanceledException oce)
+                        {
+                            LogException(e);
+                            throw new OperationCanceledException("The retry was canceled", oce);
+                        }
+
+                        retryCount++;
+                        onRetrying.OnNext(retryCount);
+                        LogRetry(retryCount);
+                    }
```

Comment:
I don't need the log output of the exception here because an OperationCanceledException is thrown.
Oops, and my implementation is wrong.
I've removed the log output!
https://github.com/extreal-dev/Extreal.Core.Common/pull/4/commits/2a52c11ef0bbfe92121ff8cb45e1000425242cd0

Reply:
確認しました！